### PR TITLE
Abstract skills

### LINKED ISSUE #362: Need ability to create "Abstract" skills
**Is your feature request related to a problem? Please describe.**
The carpark agent and weather station agent share 90% of the same code- but currently this needs to be duplicated and is a bit of a maintenance nightmare. 

In addition there is a LOT of code in these agents deal with the logic of negotiation using a FIPA protocol - rather than business logic.

**Describe the solution you'd like**
Approach 1: Ability to have abstract skills with "virtual functions" so both carpark and weather station would "inherit" from a FIPANegotation skill and just fill in their own logic. When you "scaffold" a skill, you could optionally pass in a base class.

Approach 2: Have a FipaNegotation skill which is not abstract, but is actually instantiated, however, then have a carpark skill that is able to set it up (perhaps passing in functions where bespoke logic is needed). 

**Describe alternatives you've considered**
There are probably other way to achieve this aim. I haven't looked into it much.

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Introduce 'abstract skills': skills that are not instantiated by the framework, but importable from other skills.

## Fixes

Fixes #362 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/config.md [hunk 1]
```diff
@@ -25,7 +25,8 @@ connections:                                    # The list of connection public
 contracts: []                                   # The list of contract public ids the AEA project depends on (each public id must satisfy PUBLIC_ID_REGEX).
 protocols:                                      # The list of protocol public ids the AEA project depends on (each public id must satisfy PUBLIC_ID_REGEX).
 - fetchai/default:0.3.0
-skills:                                         # The list of skill public ids the AEA project depends on (each public id must satisfy PUBLIC_ID_REGEX).
+skills: []                                      # The list of skill public ids the AEA project depends on (each public id must satisfy PUBLIC_ID_REGEX).
+is_abstract: false                              # A boolean that if `true` makes the skill abstract, i.e. not instantiated by the framework but importable from other skills. Defaults to `false`.
```
Comment (DavidMinarsch) [inline docs/config.md:29]: This should be declared as an optional field here

Comment (marcofavorito) [reply docs/config.md:29]: 'Defaults to `false`" already suggests it is optional. I will make more clear.

File: aea/aea_builder.py [hunk 2]
```diff
@@ -1300,6 +1300,11 @@ def _load_and_add_components(
             if configuration in self._component_instances[component_type].keys():
                 component = self._component_instances[component_type][configuration]
             else:
+                if component_type == ComponentType.SKILL:
+                    is_skill_abstract = cast(SkillConfig, configuration).is_abstract
+                    if is_skill_abstract:
+                        load_aea_package(configuration)
+                        continue
```
Comment (marcofavorito) [inline aea/aea_builder.py:1307]: very ugly, but I think it's the best place where to do this check. `Skill.from_config` should still return a component instance, even if abstract. Thoughts?

Comment (DavidMinarsch) [reply aea/aea_builder.py:1307]: At least extract it as a generic helper function so we can then extend it for other components, something like:
```
if ...
elif is_abstract_component(configuration):
    load_aea_package(configuration)
else:
    ....
```
where `is_abstract_component` is a helper function in helpers

**Pull request merged**